src/openrct2-ui/windows/Map.cpp

```diff
@@ -136,1448 +128,1272 @@ static constexpr const ScreenCoordsXY MiniMapOffsets[] = {
+        gMapSelectFlags &= ~MAP_SELECT_FLAG_ENABLE_ARROW;
+        auto mapCoords = footpath_bridge_get_info_from_pos(screenCoords, &direction, &tileElement);
+        if (mapCoords.IsNull())
+            return;
+
+        int32_t mapZ = tileElement->GetBaseZ();
+        if (tileElement->GetType() == TileElementType::Surface)
+        {
+            if ((tileElement->AsSurface()->GetSlope() & TILE_ELEMENT_SLOPE_ALL_CORNERS_UP) != 0)
+                mapZ += 16;
+            if (tileElement->AsSurface()->GetSlope() & TILE_ELEMENT_SLOPE_DOUBLE_HEIGHT)
+                mapZ += 16;
+        }
+
         gMapSelectFlags |= MAP_SELECT_FLAG_ENABLE;
+        gMapSelectFlags |= MAP_SELECT_FLAG_ENABLE_ARROW;
         gMapSelectType = MAP_SELECT_TYPE_FULL;
         gMapSelectPositionA = mapCoords;
-        gMapSelectPositionB = mapCoords + CoordsXY{ size, size };
+        gMapSelectPositionB = mapCoords;
+        gMapSelectArrowPosition = CoordsXYZ{ mapCoords, mapZ };
+        gMapSelectArrowDirection = direction_reverse(direction);
         map_invalidate_selection_rect();
-
-        auto landSetRightsAction = LandSetRightsAction(
-            { gMapSelectPositionA.x, gMapSelectPositionA.y, gMapSelectPositionB.x, gMapSelectPositionB.y },
-            LandSetRightSetting::SetOwnershipWithChecks, _activeTool << 4);
-        GameActions::Execute(&landSetRightsAction);
     }
-}
 
-static void WindowMapTextinput(rct_window* w, rct_widgetindex widgetIndex, char* text)
-{
-    int32_t size;
-    char* end;
+    void SetPeepSpawnToolDown(const ScreenCoordsXY& screenCoords)
+    {
+        // Verify footpath exists at location, and retrieve coordinates
+        TileElement* tileElement;
+        int32_t direction;
+        auto mapCoords = footpath_get_coordinates_from_pos(screenCoords, &direction, &tileElement);
+        if (mapCoords.IsNull())
+            return;
 
-    if (text == nullptr)
-        return;
+        int32_t mapZ = tileElement->GetBaseZ();
 
-    switch (widgetIndex)
+        auto gameAction = PlacePeepSpawnAction({ mapCoords, mapZ, static_cast<Direction>(direction) });
+        auto result = GameActions::Execute(&gameAction);
+        if (result.Error == GameActions::Status::Ok)
+        {
+            OpenRCT2::Audio::Play3D(OpenRCT2::Audio::SoundId::PlaceItem, result.Position);
+        }
+    }
+
+    void OnTextInput(rct_widgetindex widgetIndex, std::string_view text) override
     {
-        case WIDX_LAND_TOOL:
-            size = strtol(text, &end, 10);
-            if (*end == '\0')
+        if (text.empty())
+            return;
+
+        switch (widgetIndex)
+        {
+            case WIDX_LAND_TOOL:
             {
-                size = std::clamp(size, MINIMUM_TOOL_SIZE, MAXIMUM_TOOL_SIZE);
-                _landRightsToolSize = size;
-                w->Invalidate();
+                char* end;
+                int32_t size = strtol(std::string(text).c_str(), &end, 10);
+                if (*end == '\0')
```

Comment:
This pointer will point to invalid memory. You need to construct the string as a seperate item not do it all in line.
```
                std::string textStr(text);
                char* end;
                int32_t size = strtol(textStr.c_str(), &end, 10);
                if (*end == '\0')
```
same for the below as well.